projects/extension/sql/idempotent/904-semantic-catalog-search.sql

```diff
@@ -0,0 +1,291 @@
+create or replace function ai._find_relevant_obj
+( catalog_id pg_catalog.int4
+, embedding @extschema:vector@.vector
+, "limit" pg_catalog.int8 default 5
+, only_objtype pg_catalog.text default null
+, max_dist pg_catalog.float8 default null
+) returns table
+( objtype pg_catalog.text
+, objnames pg_catalog.text[]
+, objargs pg_catalog.text[]
+, classid pg_catalog.oid
+, objid pg_catalog.oid
+, objsubid pg_catalog.int4
+, description pg_catalog.text
+, dist pg_catalog.float8
+)
+as $func$
+declare
+    _dimensions pg_catalog.int4;
+    _sql pg_catalog.text;
+begin
+    _dimensions = @extschema:vector@.vector_dims(embedding);
+
+    _sql = pg_catalog.format
+    ( $sql$
+    select
+      x.objtype
+    , x.objnames
+    , x.objargs
+    , x.classid
+    , x.objid
+    , x.objsubid
+    , x.description
+    , min(x.dist) as dist
+    from
+    (
+        select
+          x.objtype
+        , x.objnames
+        , x.objargs
+        , x.classid
+        , x.objid
+        , x.objsubid
+        , x.description
+        , x.embedding operator(@extschema:vector@.<=>) ($1::@extschema:vector@.vector(%s)) as dist
+        from ai.semantic_catalog_obj_%s x
+        where pg_catalog.has_schema_privilege($2, x.objnames[1], 'usage') and
+        case x.objtype
+            when 'table' then pg_catalog.has_table_privilege($2, x.objid, 'select')
+            when 'view' then pg_catalog.has_table_privilege($2, x.objid, 'select')
+            when 'table column' then pg_catalog.has_column_privilege($2, x.objid, x.objsubid::pg_catalog.int2, 'select')
+            when 'view column' then pg_catalog.has_column_privilege($2, x.objid, x.objsubid::pg_catalog.int2, 'select')
+            when 'function' then pg_catalog.has_function_privilege($2, x.objid, 'execute')
+        end
+        %s
+        %s
+        order by dist
+        limit %L
+    ) x
+    group by
+      x.objtype
+    , x.objnames
+    , x.objargs
+    , x.classid
+    , x.objid
+    , x.objsubid
+    , x.description
+    order by min(x.dist)
+    $sql$
+    , _dimensions
+    , catalog_id
+    , case
+        when only_objtype is null then ''
+        else pg_catalog.format('and x.objtype operator(pg_catalog.=) %L', only_objtype)
+      end
+    , case
+        when max_dist is null then ''
+        else pg_catalog.format('and (x.embedding operator(@extschema:vector@.<=>) ($1::@extschema:vector@.vector(%s))) <= %s', _dimensions, max_dist)
+      end
+    , "limit"
+    );
+    -- raise log '%', _sql;
+
+    return query execute _sql using embedding, pg_catalog."current_user"();
+end;
+$func$ language plpgsql stable security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+-------------------------------------------------------------------------------
+-- find_relevant_obj
+create or replace function ai.find_relevant_obj
+( prompt pg_catalog.text
+, catalog_name pg_catalog.name default 'default'
+, "limit" pg_catalog.int8 default 5
+, only_objtype pg_catalog.text default null
```

Comment:
should this be an array of valid objtypes?

Reply:
https://github.com/timescale/pgai/pull/289/commits/6cae6d278c3be6be144cd375d40bb8fb7df5583c